### PR TITLE
Optimize recursive_sha256_ni() across compilers

### PR DESCRIPTION
**Pull request by voidxno**
Updated PR for more optimized `recursive_sha256_ni()` across compilers.

Goal: Arrange source code to help/trigger optimized compiler code.
Goal: Keep to existing structure and code style.

Changes:
- Moved `recursive_sha256_ni()` to separate `sha256_ni_rec.cpp` file
- Duplicated/optimized needed functions into `sha256_ni_rec.cpp` file
- Kept function structure, not all-in-one like PR#197
- Removed/changed code with prerequisite length 32 bytes
- Original lines commented out so easier to review
- Changed `compress_digest()` from `static` to `inline`
- Defined `optimalinline` depending on Linux/Win

Observations:
- Compilers looks to like functions structure vs. all-in-one
- Compilers did not like `static` on `compress_digest()`
- gcc12 needs `__attribute__((always_inline)) inline` to trigger boost
- VC++ needs `inline`, not `__forceinline`, to get a small boost
- About ~20% performance boost vs current master (Clang15, YMMV)
- Win/VC++ still behind, but there is a small boost

Tested:
- New recursive recursive_sha256_ni() tested manually/isolated
- Known end hash after number of iterations checked
- Tested on Linux/gcc12, Linux/Clang15, Win/VC++
- Compiled/ran new source for full mmx-node (Linux/Win)
- Ran Linux/Clang15 as fastest TimeLord a few minutes
- No peers complained, rewards received

### CODE REVIEW

File: src/sha256_ni_rec.cpp [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+#if defined(__SHA__) || defined(_WIN32)
+
+#include <immintrin.h>
+
+#ifdef _WIN32
+#include <intrin.h>
+#endif
+
+#ifdef _WIN32
+#define optimalinline inline
```
Comment (madMAx43v3r) [inline src/sha256_ni_rec.cpp:26]: Macros should be upper case, like `OPTIMALINLINE`

File: include/sha256_ni.h [hunk 2]
```diff
@@ -12,8 +12,6 @@
 
 void sha256_ni(uint8_t* out, const uint8_t* in, const uint64_t length);
 
-void recursive_sha256_ni(uint8_t* hash, const uint64_t num_iters);
```
Comment (madMAx43v3r) [inline include/sha256_ni.h:15]: You can leave it here and just include the same file in your *.cpp

**Review by madMAx43v3r — CHANGES_REQUESTED** [review]

File: src/sha256_ni_rec.cpp (hunk 1 above)
Comment (voidxno) [reply src/sha256_ni_rec.cpp]: Fixed & commited.

File: include/sha256_ni.h (hunk 2 above)
Comment (voidxno) [reply include/sha256_ni.h]: Fixed & commited.

File: src/sha256_ni_rec.cpp [hunk 3]
```diff
@@ -0,0 +1,286 @@
+/*
+ * sha256_ni_rec.cpp
+ *
+ *  Created on: May 28, 2023
+ *      Author: mad
```
Comment (madMAx43v3r) [inline src/sha256_ni_rec.cpp:5]: You can put your name here ;)
Also maybe update the creation date.

**Review by madMAx43v3r — APPROVED** [review]

**Comment by voidxno** [comment]
Thx. Testing clean Ubuntu/gcc12 build, and the Windows build (when finished). Just to be safe. No message from me = went ok.

**Pull request merged**